### PR TITLE
Improve channel updates

### PR DESCRIPTION
**Pull request by granular-ryanbonham**
Improve channel updates  to use hash of manifests to ensure we always update.

### CODE REVIEW

**Comment by mikesplain** [comment]
Cleanup of https://github.com/kubernetes/kops/pull/6829

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go [hunk 1]
```diff
@@ -41,60 +42,87 @@ var _ fi.ModelBuilder = &BootstrapChannelBuilder{}
...
-		return err
-	}
 
-	addonsYAML, err := utils.YamlMarshal(addons)
-	if err != nil {
-		return fmt.Errorf("error serializing addons yaml: %v", err)
-	}
-
-	name := b.cluster.ObjectMeta.Name + "-addons-bootstrap"
+	// manifestURL, err := a.GetManifestFullUrl()
```
Comment (rdrgmnzs) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder.go]: Any reason to leave the commented out section here?

Comment (granular-ryanbonham) [reply upup/pkg/fi/cloudup/bootstrapchannelbuilder.go]: Nope, good catch. I cleaned that up.

File: upup/pkg/fi/cloudup/bootstrapchannelbuilder.go [hunk 2]
```diff
@@ -41,60 +42,72 @@ var _ fi.ModelBuilder = &BootstrapChannelBuilder{}
...
+			return fmt.Errorf("error remapping manifest %s: %v", manifestPath, err)
+		}
+
+		rawManifest := string(manifestBytes)
+		klog.V(4).Infof("Manifest %v", rawManifest)
+
+		manifestHash, err := utils.HashString(&rawManifest)
+		klog.V(4).Infof("hash %s", manifestHash)
+		if err != nil {
+			manifestHash = ""
```
Comment (justinsb) [inline upup/pkg/fi/cloudup/bootstrapchannelbuilder.go]: Typically a good idea to warn if there's any error we're ignoring.  I'd probably return error here, as we should always be able to hash a string!

File: upup/pkg/fi/utils/hash.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"strings"
+)
+
+// HashString Takes String and returns a sha1 hash represented as a string
+func HashString(s *string) (string, error) {
+	h := sha1.New()
+	ts := strings.TrimSpace(*s)
```
Comment (justinsb) [inline upup/pkg/fi/utils/hash.go]: Trimming the string probably belongs in the caller, not a generic HashString function

File: upup/pkg/fi/utils/hash.go [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+package utils
+
+import (
+	"crypto/sha1"
+	"encoding/hex"
+	"strings"
+)
+
+// HashString Takes String and returns a sha1 hash represented as a string
+func HashString(s *string) (string, error) {
```
Comment (justinsb) [inline upup/pkg/fi/utils/hash.go]: We probably shouldn't take a string pointer - it's a ref type anyway, and we can't pass nil

File: upup/pkg/fi/utils/hash.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+	h := sha1.New()
+	ts := strings.TrimSpace(*s)
+	_, err := h.Write([]byte(ts))
+	if err != nil {
+		return "", err
+	}
+	sha := h.Sum(nil)                 // "sha" is uint8 type, encoded in base16
```
Comment (justinsb) [inline upup/pkg/fi/utils/hash.go]: Nit: I think the type is `[20]byte`, actually

**Comment by justinsb** [comment]
Looks great - thanks @granular-ryanbonham !

/approve
/lgtm

**Comment by justinsb** [comment]
I'm going to create a target to fix the golden output, as now we have hashes the golden output is going to change even more often!

**Comment by justinsb** [comment]
Needs a rebase and a run of the new streamlined script in https://github.com/kubernetes/kops/pull/7291

(It's a one-liner, so you can also just run it!)

**Pull request merged**